### ISSUE TITLE
Bad Apple Demo

### ISSUE THREAD

**Issue by AdamKlob** [atarijookie/ce-atari#172]
Bad Apple (https://fenarinarsa.com/badapple/fenarinarsa_badapple_sv2k17.zip) demo gave me a lot o headaches trying to run on CosmosEx. It crashed. A lot.

I found out that it does not like 'cold' run. If I try to run it just after I turn on cold STE+CosmosEx, it will crash, most likely after a second or two. If I leave ST running for about 15 minutes before running Bad Apple, it runs ok.

I'm unsure if it my STE or CosmosEx, but all CosmosEx test pass, on cold machine.

I hope it will be a good test case.

**Comment by ggnkua** [comment]
I think I already gave it a go on my MSTE and it ran fine. But I will repeat the test and will see if I can set up the STE as well.

**Comment by tin-nl** [comment]
Are you using an translated (aka network/fat formatted USB stick) or raw (aka sd card or atari formatted usb stick) device?

**Comment by miniupnp** [comment]
I've seen fenarinarsa last saturday, it is using a UltraSatan on his 4MB 520STE and his Mega STE
But he never tried with a cosmosex in translated mode. I will try

**Comment by ggnkua** [comment]
I just tried it on translated network drive from a stone cold MSTE. 0 issues.

**Comment by AdamKlob** [comment]
I ha ce_dd at Id 0 and sd at Id 1, and I have been running it of SD card, without translation,

**Comment by tin-nl** [comment]
@AdamKlob K, thx, I‘ll test that config and report back.

**Comment by miniupnp** [comment]
@AdamKlob If you're running from SDCard, which HD Driver are you using ?

**Comment by miniupnp** [comment]
@AdamKlob  Bad apple makes a lot of use of the blitter. Maybe your blitter has issues when cold...
as few programs use blitter, a blitter problem may well stay unnoticed

**Comment by tin-nl** [comment]
Odd - I can't get BA to run on SD card at all. It just freezes during the loading bar in the first fread() that has all demo IRQs activated (at TEXT+1006, inside the folowing trap).  Translated works fine, though.
All CE Tests (TST_FSYS/CE_TSTHD) work fine.
Interrestingly I only get ~798kB/s from Hans, Didn't we get around 1MB/s on Hans transfer sometime in the past?

Tested:
- STE 4MB TOS 1.06 
- MSTe 4MB TOS 2.05
- HDDRIVER 10.03/10.10/9.03
- MicroCosmos w/o Pi
- CE V1 Yocto (latest Release)
- other CE V1 Raspbian (ce_app build from master)
- different SD cards (used ones as well as fresh partitioned)
I tried all combinations, but not all cards with all hddrivers. Files were checked after copying.

Titles that work:
- WeWere@ (HD Loader)
- Cernit Dendrafir
- Summer Delight
- Drone 

Titles that dont:
- Bad Apple (hangs at loading bar)
- Sea of colors (runs somewhat but shows garbled fx)
- Cyg/PPeras Videoplayer Moleman2.cav (has Timing issues)

BA/Seao worked before on these machines and at least both CEV1.
BA/Sea work on translated, Didn't check the Vidoplayer, doesnt use Gemdos anyway.

Can anyone recheck this so we can see if it's only a regression on my hardware/setup/testing or on Hans itself?

@AdamKlob Do Sea of Colors and P158DR.TTP (with moleman2.cav) work for you?

**Comment by ggnkua** [comment]
Okay, I just tried running the demo from hddriver 9.03 and I also get a freeze in the first fread.

It's a bit tricky to try and debug on my MSTE as there are a few bent vectors that cause all sorts of havoc with Bugaboo. Perhaps I'll set up my STE and try there, that's usually more reliable.

In my mind there's no doubt that the problem is caused exactly from those bent vectors, especially since they don't chain call the system vectors after they're done. As an experiment I tried calling the system VBL after the installed VBL routine is done but that didn't help at all. The author (fenarinarsa) promised to release the source code so it might be better to wait for this, otherwise easyrider it'll be.

Of course all this doesn't explain how it works with ultrasatan for people in the first place - isn't cosmosex supposed to be identical with an ultrasatan when it boots from sd? Quite weird. I'll try running the demo from an ultrasatan just to be sure...

**Comment by ggnkua** [comment]
Yup, I just had to test it on Ultrasatan on the same SD I used on the cosmosex - runs totally fine there. Now I'm confused.

**Comment by AdamKlob** [comment]
I was using ppera driver.

śr., 13.12.2017, 18:43 użytkownik ggnkua <notifications@github.com> napisał:

**Comment by miniupnp** [comment]
> Of course all this doesn't explain how it works with ultrasatan for people in the first place - isn't cosmosex supposed to be identical with an ultrasatan when it boots from sd? Quite weird. I'll try running the demo from an ultrasatan just to be sure...

It may be just a problem with the CosmosEx driver loaded in memory badly interacting with the HD driver.
I think fenarinarsa told me Bad Apple does all the display job in the HBL interrupt.

**Comment by tin-nl** [comment]
@AdamKlob Ah, okay, so I‘ll test that one too, then. Which CE do you use? V1, V2 or Microcosmos? Is the latest Firmware installed or an earlier version? If an earlier version: which date is shown?

@ggnkua Thanks for checking on your machine. IIRC Hans code is heavily based on Ultrasatan, but changed quite a bit this/last year due to some timeout issues with certain SD cards or sth like that. Sourcecode to BA would be indeed swell - had to switch to mono with ubeswitch all the time to see anything in Bugaboo, so I gave up, too. 
Can you give Sea of Colours and P128DR.TTP a spin too, when you‘re at it?

@miniupnp At least in my test cases the results were the same with or w/o CE_DD loaded.

**Comment by AdamKlob** [comment]
V2, latest firmware. I have V1 too and I can try on it too.

śr., 13.12.2017, 20:43 użytkownik tin-nl <notifications@github.com> napisał:

**Comment by miniupnp** [comment]
just ran the bad apple demo fine on my STE with MicroCosmos...

**Comment by fenarinarsa** [comment]
Hi, I know it might crash in some cases, I only used Pperas drivers with UltraSatan :) The only incompatibility I found is when BIGDOS.PRG is loaded in memory, it always crashes.

I didn't test HDDRIVER since I don't have it. Also I couldn't test ACSI access on emulators, when I install Ppera's driver on Hatari with a blank ACSI image, it hangs on the loading screen (I don't know why). Since it worked on both my MSTE and STE I deduced it was an issue with the emulator.

**Comment by fenarinarsa** [comment]
(note that I will release the source code, I just currently don't have time to clean it up and make a correct archive with all needed assets)

**Comment by tin-nl** [comment]
BA works with Ppera driver on a warm machine. Retested on same partition with Hddriver 9.x and 10.10 - freezes at the loading bar. Sea of Color also works with pperas driver, but doesn‘t with Hddriver. Have to check on a cold machine.

 But this probably isn‘t a sole Hddriver issue - SoC used to work with Hddriver 9.x on an older CE firmware. P158DR.TTP/Moleman2 also did work on an older CE firmware. Maybe it‘s possible to downgrade the Hans firmware (should be enough, I think) to confirm this.

@fenarinarsa: congrats for the demo - quite bold to stream video in Gemdos and not using an own built-in driver, like others do :)

**Comment by tin-nl** [comment]
Cold run of BA with Pperas driver works here.
